### ISSUE TITLE
[Security] global ssh_config is not honored

### ISSUE THREAD

**Issue by JW0914** [PowerShell/Win32-OpenSSH#847]
**Version:** 0.0.18.0
**Server:** LEDE  (Linux router OS) 
**Client:** Windows 10 Pro

Win32-OpenSSH refuses to honor the `Ciphers` option when set in ssh_config *(likely sshd_config as well)*.
- ***C:\Program Files\OpenSSH\ssh_config:*** 
   `Ciphers                           = rijndael-cbc@lysator.liu.se,aes256-cbc,aes192-cbc,aes128-cbc`
- ***/etc/ssh/sshd_config:*** 
   `Ciphers                           = rijndael-cbc@lysator.liu.se,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc`

This results with:
-  **Client:** *Win32-OpenSSH error*
   ```
   Unable to negotiate with <host> port <#>: no matching cipher found. Their offer: aes128-cbc,aes192-cbc,aes256-cbc
   ```
- **Server:** *LEDE OpenSSH [`sshd`] error*
   ```
   Unable to negotiate with <host> port <#>: no matching cipher found. Their offer: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr [preauth]
   ```
- **Available Server / Client Ciphers:**

[root@LEDE] ~ # ssh -Q cipher | PS C:\Users\JW0914> ssh -Q cipher
---------------------------------- | --------------------------------------- 
3des-cbc | 3des-cbc
blowfish-cbc | -
cast128-cbc | -
arcfour | -
arcfour128 | -
arcfour256 | -
aes128-cbc | -
aes192-cbc | aes192-cbc
aes256-cbc | aes256-cbc
rijndael-cbc@lysator.liu.se | rijndael-cbc@lysator.liu.se
aes128-ctr | aes128-ctr
aes192-ctr | aes192-ctr
aes256-ctr | aes256-ctr
aes128-gcm@openssh.com | -
aes256-gcm@openssh.com | -
chacha20-poly1305@openssh.com | chacha20-poly1305@openssh.com
- **[root@LEDE] ~ # logread -f | grep ssh**
   ```
   00:21:26 debug1: Forked child 16328.
   00:21:26 debug1: Set /proc/self/oom_score_adj to 0
   00:21:26 debug1: rexec start in 6 out 6 newsock 6 pipe 8 sock 9
   00:21:26 debug1: inetd sockets after dupping: 5, 5
   00:21:26 auth.info: Connection from <host> port <#> <host> port <#>
   00:21:26 debug1: Client protocol version 2.0; client software version OpenSSH_7.5
   00:21:26 debug1: match: OpenSSH_7.5 pat OpenSSH* compat 0x04000000
   00:21:26 debug1: Local version string SSH-2.0-OpenSSH_7.5
   00:21:26 debug1: Enabling compatibility mode for protocol 2.0
   00:21:26 debug2: fd 5 setting O_NONBLOCK
   00:21:26 debug2: Network child is on pid 16359
   00:21:26 debug1: permanently_set_uid: 22/22 [preauth]
   00:21:26 debug1: ssh_sandbox_child: prctl(PR_SET_SECCOMP): Invalid argument [preauth]
   00:21:26 debug1: list_hostkey_types: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ssh-ed25519 [preauth]
   00:21:26 debug1: SSH2_MSG_KEXINIT sent [preauth]
   00:21:26 debug1: SSH2_MSG_KEXINIT received [preauth]
   00:21:26 debug2: local server KEXINIT proposal [preauth]
   00:21:26 debug2: KEX algorithms: curve25519-sha256@libssh.org,curve25519-sha256,diffie-hellman-group-exchange-sha256,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521 [preauth]
   00:21:26 debug2: host key algorithms: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ssh-ed25519 [preauth]
   00:21:26 debug2: ciphers ctos: rijndael-cbc@lysator.liu.se,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc [preauth]
   00:21:26 debug2: ciphers stoc: rijndael-cbc@lysator.liu.se,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc [preauth]
   00:21:26 debug2: MACs ctos: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha2-256,hmac-sha2-512 [preauth]
   00:21:26 debug2: MACs stoc: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha2-256,hmac-sha2-512 [preauth]
   00:21:26 debug2: compression ctos: none,zlib@openssh.com [preauth]
   00:21:26 debug2: compression stoc: none,zlib@openssh.com [preauth]
   00:21:26 debug2: languages ctos:  [preauth]
   00:21:26 debug2: languages stoc:  [preauth]
   00:21:26 debug2: first_kex_follows 0  [preauth]
   00:21:26 debug2: reserved 0  [preauth]
   00:21:26 debug2: peer client KEXINIT proposal [preauth]
   00:21:26 debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,ext-info-c [preauth]
   00:21:26 debug2: host key algorithms: ssh-ed25519-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256,ssh-rsa [preauth]
   00:21:26 debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr [preauth]
   00:21:26 debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr [preauth]
   00:21:26 debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [preauth]
   00:21:26 debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [preauth]
   00:21:26 debug2: compression ctos: none [preauth]
   00:21:26 debug2: compression stoc: none [preauth]
   00:21:26 debug2: languages ctos:  [preauth]
   00:21:26 debug2: languages stoc:  [preauth]
   00:21:26 debug2: first_kex_follows 0  [preauth]
   00:21:26 debug2: reserved 0  [preauth]
   00:21:26 debug1: kex: algorithm: curve25519-sha256 [preauth]
   00:21:26 debug1: kex: host key algorithm: ssh-ed25519 [preauth]
   00:21:26 auth.info: Unable to negotiate with <host> port <#>: no matching cipher found. Their offer: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr [preauth]
   00:21:26 debug1: do_cleanup [preauth]
   00:21:26 debug1: monitor_read_log: child log fd closed
   00:21:26 debug1: do_cleanup
   00:21:26 debug1: Killing privsep child 16359
   ```

[**sshd_config** *(LEDE)*.txt](https://github.com/PowerShell/Win32-OpenSSH/files/1233367/sshd_config.LEDE.txt) | [**ssh_config** *(Win32-OpenSSH)*.txt](https://github.com/PowerShell/Win32-OpenSSH/files/1233348/ssh_config.Win32-OpenSSH.txt) | [**sshd_config** *(Win32-OpenSSH)*.txt](https://github.com/PowerShell/Win32-OpenSSH/files/1233349/sshd_config.Win32-OpenSSH.txt)

**Comment by JW0914** [comment]
The issue appears to be `ssh.exe` refuses to acknowledge the *ssh_config* file stored in *C:\Program Files\OpenSSH\\* (unless specifically referenced via `-F 'C:\Program Files\OpenSSH\ssh_config'`), instead only recognizing an ssh_config stored at *%USERPROFILE%\\.ssh\\config* by default.

- This is **not** normal behavior for OpenSSH and should this not be fixed, it needs to be annotated in the [ReadMe](https://github.com/PowerShell/Win32-OpenSSH/blob/L1-Prod/README), as generic OpenSSH information in the ReadMe is not only not helpful, it's irrelevant to this repo.
  - Normal behavior takes system wide settings from *OpenSSH\ssh_config* and user specific/custom settings from *.ssh\config*

**Comment by manojampalam** [comment]
Added as security considerations in Wiki and put a link [here](https://github.com/PowerShell/Win32-OpenSSH/wiki/Install-Win32-OpenSSH). 
Related to #466.

**Comment by JW0914** [comment]
I appreciate this being annotated under [Security Considerations](https://github.com/PowerShell/Win32-OpenSSH/wiki/Security-Considerations), however that doesn't fully address the issue...

Currently, Win32-OpenSSH does not recognize _anything_ set in **C:\Program Files\OpenSSH\ssh_config** unless the `-F` switch is utilized (due entirely to how it's currently compiled), and while that's a massive security flaw in and of itself, there are many non-security related functions set via the system wide ssh_config.

This information should not be buried several links deep in the repo, and at the very least, should be in line on the [installation page](https://github.com/PowerShell/Win32-OpenSSH/wiki/Install-Win32-OpenSSH), or better yet, have the installation script create **%USERPROFILE%\\.ssh\\config**.
- Simply put, the contributors to this repo are shooting themselves, and Microsoft, in the foot if this is not annotated front and center because it is such a vital part of how OpenSSH functions. 
  - _How many will simply abandon trying to utilize Win32-OpenSSH, when they're unable to figure out how to work around the system wide client ssh_config not being recognized by `ssh.exe`?_  
- **%USERPROFILE%\\.ssh\\config** is not meant to function as the system wide ssh_config, but adjacent to.

**Comment by manojampalam** [comment]
Thanks for detailing the impact. We'll put in a Windows specific block in code to accommodate this request for now.

**Comment by JW0914** [comment]
@manojampalam Thank you, I appreciate that.

**Comment by manojampalam** [comment]
Reopening to keep track of this for the upcoming milestone.

**Comment by bingbing8** [comment]
fix as part of https://github.com/PowerShell/openssh-portable/pull/258
